File: level1/lib/channels/telegram_channel.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart' show MediaType;

import '../core/channel.dart';
import '../core/models.dart';

class TelegramChannelSettings implements ChannelSettings {
  final String botToken;
  final String chatId;

  const TelegramChannelSettings({required this.botToken, required this.chatId});

  @override
  String get type => 'telegram';

  @override
  Map<String, dynamic> toJson() => {'botToken': botToken, 'chatId': chatId};

  @override
  List<String> get secretFields => ['botToken'];

  factory TelegramChannelSettings.fromJson(Map<String, dynamic> json) =>
      TelegramChannelSettings(
        botToken: json['botToken'] as String? ?? '',
        chatId: json['chatId'] as String? ?? '',
      );
}

class TelegramChannel extends Channel {
  @override
  final String id;
  @override
  final bool enabled;
  @override
  final TelegramChannelSettings settings;
  final http.Client _client;
  static final _tokenRe = RegExp(r'^\d+:[A-Za-z0-9_-]+$');

  TelegramChannel({
    required this.id,
    this.enabled = true,
    required this.settings,
    http.Client? client,
  }) : _client = client ?? http.Client();

  @override
  String get type => 'telegram';

  static String _endpoint(String method, String token) =>
      'https://api.telegram.org/bot$token/$method';

  @override
  Future<void> send(AlertMessage message) async {
    final token = settings.botToken;
    if (message.snapshot != null) {
      final photo = message.snapshot!;
      final photoOk = await _sendPhoto(token, settings.chatId, photo, message.text);
      if (!photoOk) {
        await _sendMessage(token, settings.chatId, message.text);
      }
    } else {
      await _sendMessage(token, settings.chatId, message.text);
    }
  }

  Future<bool> _sendPhoto(
      String token, String chatId, Snapshot photo, String caption) async {
    final request = http.MultipartRequest('POST', Uri.parse(_endpoint('sendPhoto', token)))
      ..fields['chat_id'] = chatId
      ..fields['caption'] = caption
      ..files.add(http.MultipartFile.fromBytes(
        'photo',
        photo.bytes,
        filename: photo.name,
        contentType: MediaType.parse(photo.mimeType),
      ));
    final streamed = await _client.send(request);
    final response = await http.Response.fromStream(streamed);
    return _ok(response);
  }

  Future<void> _sendMessage(String token, String chatId, String text) async {
    final response = await _client.post(
      Uri.parse(_endpoint('sendMessage', token)),
      headers: {'content-type': 'application/json'},
      body: jsonEncode({'chat_id': chatId, 'text': text}),
    );
    if (!_ok(response)) {
      throw StateError('Telegram sendMessage failed (${response.statusCode})');
    }
  }

  bool _ok(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) return false;
    try {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      return body['ok'] == true;
    } catch (_) {
      return false;
    }
  }

  @override
  Future<void> sendTest() async {
    await _sendMessage(settings.botToken, settings.chatId, 'Security Cam: test alert');
  }

  @override
  String? validate() {
    if (settings.botToken.isEmpty || settings.chatId.isEmpty) {
      return 'Bot token and chat ID are required';
    }
    if (!_tokenRe.hasMatch(settings.botToken)) {
      return 'Bot token is not in the expected format';
    }
    return null;
  }
}

File: level2/android/app/src/main/kotlin/io/securitycam/level2/channels/TelegramChannel.kt
package io.securitycam.level2.channels

import io.securitycam.level2.core.AlertMessage
import io.securitycam.level2.core.ChannelSettings
import io.securitycam.level2.core.Snapshot
import java.util.concurrent.TimeUnit
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File

class TelegramChannelSettings(
    val botToken: String = "",
    val chatId: String = "",
) : ChannelSettings() {
    override val type: String get() = "telegram"
    override fun toJson(): Map<String, Any?> = mapOf("botToken" to botToken, "chatId" to chatId)
    override val secretFields: List<String> get() = listOf("botToken")

    companion object {
        fun fromJson(json: Map<String, Any?>): TelegramChannelSettings = TelegramChannelSettings(
            botToken = json["botToken"] as? String ?: "",
            chatId = json["chatId"] as? String ?: "",
        )
    }
}

/** Sends alerts to the Telegram Bot API (sendPhoto with sendMessage fallback). */
class TelegramChannel(
    override val id: String,
    override val enabled: Boolean = true,
    override val settings: TelegramChannelSettings,
    client: OkHttpClient? = null,
) : io.securitycam.level2.core.Channel {

    private val client: OkHttpClient =
        client ?: OkHttpClient.Builder().callTimeout(30, TimeUnit.SECONDS).build()

    override val type: String get() = "telegram"

    private fun endpoint(method: String): String =
        "https://api.telegram.org/bot${settings.botToken}/$method"

    override suspend fun send(message: AlertMessage) {
        val photo = message.snapshot
        if (photo != null) {
            if (!sendPhoto(photo, message.text)) {
                sendMessage(message.text)
            }
        } else {
            sendMessage(message.text)
        }
    }

    private suspend fun sendPhoto(photo: Snapshot, caption: String): Boolean {
        val tmp = File.createTempFile("level2", ".img")
        try {
            tmp.writeBytes(photo.bytes)
            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("chat_id", settings.chatId)
                .addFormDataPart("caption", caption)
                .addFormDataPart(
                    "photo",
                    photo.name,
                    tmp.asRequestBody(photo.mimeType.toMediaType()),
                )
                .build()
            val response = client.newCall(Request.Builder().url(endpoint("sendPhoto")).post(body).build()).execute()
            response.use { return isOk(it.body?.string()) }
        } finally {
            tmp.delete()
        }
    }

    private suspend fun sendMessage(text: String) {
        val body = jsonEncode("chat_id" to settings.chatId, "text" to text)
            .toRequestBody("application/json".toMediaType())
        val response = client.newCall(Request.Builder().url(endpoint("sendMessage")).post(body).build()).execute()
        response.use {
            if (!isOk(it.body?.string())) {
                error("Telegram sendMessage failed (${it.code})")
            }
        }
    }

    private fun isOk(body: String?): Boolean {
        if (body == null) return false
        // Telegram replies {"ok":true,...}; a targeted match avoids pulling a
        // JSON parser into the unit-test classpath.
        return OK_REGEX.containsMatchIn(body)
    }

    override suspend fun sendTest() {
        sendMessage("Security Cam: test alert")
    }

    override fun validate(): String? {
        if (settings.botToken.isEmpty() || settings.chatId.isEmpty()) {
            return "Bot token and chat ID are required"
        }
        if (!TOKEN_REGEX.matches(settings.botToken)) {
            return "Bot token is not in the expected format"
        }
        return null
    }

    companion object {
        // \d+:[A-Za-z0-9_-]+
        private val TOKEN_REGEX = Regex("^\\d+:[A-Za-z0-9_-]+$")
        private val OK_REGEX = Regex("\"ok\"\\s*:\\s*true")
    }
}